br/pkg/utils/register.go

```diff
@@ -0,0 +1,280 @@
+	defer tr.wg.Done()
+	const minTimeLeftThreshold time.Duration = 20 * time.Second
+	var timeLeftThreshold time.Duration = tr.ttl / 4
+	if timeLeftThreshold < minTimeLeftThreshold {
+		timeLeftThreshold = minTimeLeftThreshold
+	}
+	failpoint.Inject("brie-task-register-always-grant", func(_ failpoint.Value) {
+		timeLeftThreshold = tr.ttl
+	})
+	var lastUpdateTime time.Time = time.Now()
+	var err error
+	for {
+	CONSUMERESP:
+		for {
+			failpoint.Inject("brie-task-register-keepalive-stop", func(_ failpoint.Value) {
+				if _, err = tr.client.Lease.Revoke(ctx, tr.curLeaseID); err != nil {
+					log.Warn("brie-task-register-keepalive-stop", zap.Error(err))
+				}
+			})
+			select {
+			case <-ctx.Done():
+				return
+			case _, ok := <-ch:
+				if !ok {
+					break CONSUMERESP
+				}
+				lastUpdateTime = time.Now()
+			}
+		}
+		log.Warn("the keepalive channel is closed, try to recreate it")
+		needReputKV := false
+	RECREATE:
+		for {
+			timeGap := time.Since(lastUpdateTime)
+			if tr.ttl-timeGap <= timeLeftThreshold {
+				lease, err := tr.grant(ctx)
+				failpoint.Inject("brie-task-register-failed-to-grant", func(_ failpoint.Value) {
+					err = errors.New("failpoint-error")
+				})
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+					}
+					log.Warn("failed to grant lease", zap.Error(err))
+					time.Sleep(RegisterRetryInternal)
+					continue
+				}
+				tr.curLeaseID = lease.ID
+				lastUpdateTime = time.Now()
+				needReputKV = true
+			}
+			if needReputKV {
+				// if the lease has expired, need to put the key again
+				_, err := tr.client.KV.Put(ctx, tr.key, "", clientv3.WithLease(tr.curLeaseID))
+				failpoint.Inject("brie-task-register-failed-to-reput", func(_ failpoint.Value) {
+					err = errors.New("failpoint-error")
+				})
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+					}
+					log.Warn("failed to put new kv", zap.Error(err))
+					time.Sleep(RegisterRetryInternal)
+					continue
+				}
+				needReputKV = false
+			}
+			// recreate keepalive
+			ch, err = tr.client.Lease.KeepAlive(ctx, tr.curLeaseID)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				log.Warn("failed to create new kv", zap.Error(err))
+				time.Sleep(RegisterRetryInternal)
+				continue
+			}
+
+			break RECREATE
+		}
+	}
+}
+
+// RegisterTask saves the task's information
+type RegisterTask struct {
+	Key     string
+	LeaseID int64
+	TTL     int64
+}
+
+// MessageToUser marshal the task to user message
+func (task RegisterTask) MessageToUser() string {
+	return fmt.Sprintf("[ key: %s, lease-id: %x, ttl: %ds ]", task.Key, task.LeaseID, task.TTL)
+}
+
+type RegisterTasksList struct {
+	Tasks []RegisterTask
+}
+
+func (list RegisterTasksList) MessageToUser() string {
+	var tasksMsgBuf strings.Builder
+	for _, task := range list.Tasks {
+		tasksMsgBuf.WriteString(task.MessageToUser())
```

Comment:
Should we add some whitespaces between tasks?

Reply:
done